### PR TITLE
Clarify `==` docs for collections

### PR DESCRIPTION
**Pull request by mbauman**
Alright, here's my spitball at fixing https://github.com/JuliaLang/julia/issues/52484.

This also clarifies the non-missing behaviors, and adds a reference to `all` which should be a useful breadcrumb.  I don't want to say `all(x .== y)` because that has all sorts of broadcasting semantics, but we could perhaps add a comparison to `all(splat(==), zip(x, y))` or `all(map(==, x, y))`, but even those feel sketchy to reference as they may miss other properties (like array shape).

### CODE REVIEW

**Comment by Seelengrab** [comment]
What does `pairs` mean for e.g. comparing two `Set`? Consider:

```
julia> Set([missing, 1]) == Set([missing, 2])
false

julia> Set([missing, 1]) == Set([missing, 1])
true
```

**Comment by mbauman** [comment]
This doesn’t talk about `Base.pairs` and there is an out for collections to do something different with the “generally”s.

**Comment by Seelengrab** [comment]
I think it would be better to prescribe exact required properties of `==` instead of giving "generally, but not always" applicable advice. Ambiguities in docstrings are much more likely to lead to subtle bugs & confusions than what is gained from being welcome to newcomers.

**Comment by mbauman** [comment]
My goal is to be as descriptive as is possible, when individual collections (and more broadly, arbitrary types) can and **do** define semantics that make sense for themselves.  For example, `Ref` is iterable (so ostensibly could be called a "collection" of one element), but its reason-for-being is about mutable semantics — so not even `Ref(1) == Ref(1)`.  Or the fact that ordering doesn't matter for `AbstractSet`, so those are compared not by lock-step sequential iteration, but rather by `issubset`.

On the prescriptive side, it's all about the "abstract value" that the types themselves represent.  Yes, that's a hard thing to nail down, and yes, that's pretty vague:

> Should be implemented for all types with a notion of equality, based on the abstract value that an instance represents

**Review by nalimilan — APPROVED** [review]
Looks like a strict improvement over the current docstring to me. I don't see how we could give a more precise documentation given that various types follow different behaviors. `==(::AbstractSet, ::AbstractSet)` should have a separate docstring IMO, as currently it's hard or impossible to find out how it behaves, and it clearly doesn't fit in the general docstring.

**Review by simeonschaub — APPROVED** [review]
SGTM

**Comment by Seelengrab** [comment]
> I don't see how we could give a more precise documentation given that various types follow different behaviors.

My problem is that the current wording/layout mixes in the "oh and you can take this nebulous thing as _guidance_" with "here are the hard requirements of `==` you MUST obey". This makes it seem like they are equally important when implementing `==` on a custom type, which is not the case - the first is an actual, required invariant, the latter is a judgement call.

**Comment by mbauman** [comment]
@Seelengrab do you have any concrete suggestions here? My goal is to describe the status quo, warts and all, in a way that makes sense to both users and devs.

**Comment by Seelengrab** [comment]
At the moment, there are too many disparate things documented in the generic `==` fallback docstring, in an confusing order. We have 

 1. It's an equality operator
 2. Generic/"if you feel it applies" style advice
 3. Behavior for floating point types from Base
 4. Requirements for behavior on `missing` input
 5. A remark about when to use `isequal`

This makes the docstring hard to parse, understand & implement correctly; a reader might only read up to the mention of floating point types, see that their type is not a floating point type based on IEEE-745 and stop reading. I'd suggest moving the actual implementation requirement for `==` up top, followed by when to use (& implement!) `isequal` instead, followed by how floating point types should behave (or move that to a docstring documenting methods that handle those) and only then talk about more generic situations that may or may not apply.

This should make the flow of information about what is guaranteed & what is required clearer, by first emphasizing common behavior that ALL implementations must obey and only then moving to more specific situations that may or may not apply (though generally, too much non-guaranteed prose should IMO be moved to the manual entirely, and replaced with a reference to the manual in the docstring).

**Comment by mbauman** [comment]
Ok, those all seems to be larger scale changes.  This is just trying to remove inaccuracies and clarify the docstring for collections and directly address #52484.

File: base/operators.jl [hunk 1]
```diff
@@ -52,17 +52,18 @@ Generic equality operator. Falls back to [`===`](@ref).
 Should be implemented for all types with a notion of equality, based on the abstract value
 that an instance represents. For example, all numeric types are compared by numeric value,
 ignoring type. Strings are compared as sequences of characters, ignoring encoding.
-For collections, `==` is generally called recursively on all contents,
-though other properties (like the shape for arrays) may also be taken into account.
+Collections of the same type generally iterate through both operands and compare each pair
+of values, returning true if all such pairs are `==`, though other properties may also be
```
Comment (vtjnash) [inline base/operators.jl:56]:
```suggestion
Collections of the same type generally compare their key sets, and if those are `==`, then compare the values 
for each of those keys, returning true if all such pairs are `==`. Other properties are typically not taken into account (such as the exact type)
```

**Review by vtjnash — COMMENTED** [review]
Here's my attempt at improving this suggestion. I felt that `iteration` sounded too much like demanding the ordering matters for `==`. In this change, I am attempting to reflect that `==` is more closely associated with the `getindex` API than the `iterate` API.

File: base/operators.jl (hunk 1 above)
Comment (Seelengrab) [reply base/operators.jl:56]: For collections alone, this reads much better, yes.

**Comment by Seelengrab** [comment]
> Ok, those all seem to be larger scale changes.

Yeah, it's probably a bit more finicky than just moving the bits of the current docstring around - I do want to make that "shuffling" PR, but I'll hold off on that until the particular bits about collections is clarified so we don't have two PRs working on the same text at the same time.

File: base/operators.jl [hunk 2]
```diff
@@ -52,17 +52,18 @@ Generic equality operator. Falls back to [`===`](@ref).
 Should be implemented for all types with a notion of equality, based on the abstract value
 that an instance represents. For example, all numeric types are compared by numeric value,
 ignoring type. Strings are compared as sequences of characters, ignoring encoding.
-For collections, `==` is generally called recursively on all contents,
-though other properties (like the shape for arrays) may also be taken into account.
+Collections of the same type generally compare their key sets, and if those are `==`, then compare the values 
+for each of those keys, returning true if all such pairs are `==`. Other properties are typically not taken into account (such as the exact type)
+taken into account (like the shape for `AbstractArray`s).
```
Comment (mbauman) [inline base/operators.jl:57]:
```suggestion
for each of those keys, returning true if all such pairs are `==`.
Other properties are typically not taken into account (such as the exact type).
```

File: base/operators.jl [hunk 3]
```diff
@@ -52,17 +52,18 @@ Generic equality operator. Falls back to [`===`](@ref).
 Should be implemented for all types with a notion of equality, based on the abstract value
 that an instance represents. For example, all numeric types are compared by numeric value,
 ignoring type. Strings are compared as sequences of characters, ignoring encoding.
-For collections, `==` is generally called recursively on all contents,
-though other properties (like the shape for arrays) may also be taken into account.
+Collections of the same type generally compare their key sets, and if those are `==`, then compare the values
```
Comment (inkydragon) [inline base/operators.jl:55]:
```suggestion
Collections of the same type generally compare their key sets, and if those are `==`, then compare the values
```

**Review by fingolfin — APPROVED** [review]
